### PR TITLE
Mttl 771 add contact details endpoint

### PR DESCRIPTION
**Pull request by evak6979**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/MTTL-771

## Describe this PR

Added endpoint to create contact

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs [hunk 1]
```diff
@@ -70,5 +77,37 @@ public void GetContactDetailsByTargetIdThrowsException()
 
             func.Should().Throw<ApplicationException>().WithMessage(exception.Message);
         }
+
```
Comment (humulla) [inline ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs:80]: Maybe we should add a not found test here

File: ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            apiResult.Should().BeEquivalentTo(expected);
+        }
+
+        private async Task<ContactDetailsResponseObject> ExtractResultFromHttpResponse(HttpResponseMessage response)
+        {
+            response.StatusCode.Should().Be(HttpStatusCode.Created);
+            var responseContent = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            var apiResult = JsonSerializer.Deserialize<ContactDetailsResponseObject>(responseContent, _jsonOptions);
+
+            return apiResult;
```
Comment (humulla) [inline ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs:55]: Would you not need to add method for when a bad request and not found response happens?

**Review by humulla — COMMENTED** [review]
Looks good so far, just a couple of comments on the tests

File: ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs (hunk 1 above)
Comment (evak6979) [reply ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs:80]: I'll add one here yes, good call.

File: ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs (hunk 2 above)
Comment (evak6979) [reply ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs:55]: I think not found/bad request is a bit of an overkill on E2E. I'll do that in unit tests. Thanks :)

File: ContactDetailsApi/V1/Gateways/DynamoDbGateway.cs [hunk 3]
```diff
@@ -44,5 +44,15 @@ public async Task<List<ContactDetails>> GetContactDetailsByTargetId(ContactQuery
 
             return contactDetailsEntities.ToDomain();
         }
+
+        [LogCall]
+        public async Task<ContactDetails> CreateContact(ContactDetailsRequestObject requestObject)
+        {
+            var personDbEntity = requestObject.ToDatabase();
```
Comment (mikemoody-amido) [inline ContactDetailsApi/V1/Gateways/DynamoDbGateway.cs:51]: Surely a `contact` not a `personDbEntity`?

File: ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+            _lastResponse = await _httpClient.SendAsync(message).ConfigureAwait(false);
+        }
+
+        public async Task ThenTheContactDetailsAreReturned(ContactDetailsRequestObject expectedContacts)
+        {
+            var expected = expectedContacts;
+
+            var apiResult = await ExtractResultFromHttpResponse(_lastResponse).ConfigureAwait(false);
+            apiResult.Should().BeEquivalentTo(expected);
```
Comment (mikemoody-amido) [inline ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs:46]: Should we not also validate that there is actually a corresponding record in the database?

File: ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+using System;
+using ContactDetailsApi.V1.Domain;
+
+namespace ContactDetailsApi.V1.Boundary.Request
+{
+    public class ContactDetailsRequestObject
```
Comment (mikemoody-amido) [inline ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs:6]: You need to add a validator class/classes (and corresponding unit tests) for this request class.

File: ContactDetailsApi/V1/Gateways/DynamoDbGateway.cs (hunk 3 above)
Comment (evak6979) [reply ContactDetailsApi/V1/Gateways/DynamoDbGateway.cs:51]: Surely !

File: ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs (hunk 4 above)
Comment (evak6979) [reply ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs:46]: I'll add it.

File: ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs (hunk 5 above)
Comment (evak6979) [reply ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs:6]: Validation is a separate ticket I think.

File: ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs (hunk 1 above)
Comment (evak6979) [reply ContactDetailsApi.Tests/V1/Controllers/ContactDetailsControllerTests.cs:80]: I am not sure what not found would mean on the basis of a contact actually?

File: ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs (hunk 5 above)
Comment (mikemoody-amido) [reply ContactDetailsApi/V1/Boundary/Request/ContactDetailsRequestObject.cs:6]: Cool. We can do it there then.

File: ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs (hunk 4 above)
Comment (mikemoody-amido) [reply ContactDetailsApi.Tests/V1/E2ETests/Steps/CreateContactSteps.cs:55]: I would have just gone straight to the database, but what you've done works too. :-)

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**